### DISCUSSION TITLE
What is useState?

### DISCUSSION THREAD

**Discussion by honaseung** [honaseung/react-deep-dive#21] [Chapter3]
If you are familiar with `class component`, then you know `state`. Are you familiar but don't know? You are lying to me. `useState` is a function which make a state from `function component`. Then what is `state`? Do you know exactly? If you can tell my 3 questions below. You already know `state` perfectly. If not, sit down and read this discussion.

 1. Why you should use state?
 2. How state work?
 3. Which one is good for make as state?

I think these are most simple and basical question when you talk about `state`. How many answers do you have? No perfect? Come on! Here we go from number1.

### 1. Why you should use state?
> In React, React render components many times. When React re-render it, It means It will start to make components once again. Let's  guess your components goes like below,
```
const Test = () => {
  const a = 1;
  return (
    <div>TEST</div>
  );
};
```

> When React re-render this, The codes are ran start from `const a = 1;`.  Now we get it. Don't you? We can't keep a's value. Even if you change it before re-render `Test`. a's value is always 1. **If you want to make React functional component remember the changed value (or previous value too). You should use `useState`.**

### 2. How state work?
> We talked about 'What is state'. This time we're gonna talk about how it work. At real raw level, It might be (Yes I'm guessing now. I can't know how it work at real raw level.) like below.
```
At the raw level
.
.
.
At first render
const [stateA, setStateA] = useStateA(0); => stateA = 0;
At re-render after setStateA(1);
const [stateA, setStateA] = useStateA(1); => stateA = 1;
```
> It will be just like you assign variable at which render. It means every render could have different `state`. One of the reason why render happened is `setState` as you know. **So if state is changed, re-render is happened with new `state`.** If a component has states rather than state, then how it goes on? It should go like below.
```
At first render
const [stateA, setStateA] = useStateA(0); => stateA = 0;
const [stateB, setStateB] = useStateB(0); => stateB = 0;
const [stateC, setStateC] = useStateC(0); => stateC = 0;
At re-render after setStateA(1);
const [stateA, setStateA] = useStateA(1); => stateA = 1;
const [stateB, setStateB] = useStateB(0); => stateB = 0;
const [stateC, setStateC] = useStateC(0); => stateC = 0;
At re-render after setStateA(2) and setStateC(3);
const [stateA, setStateA] = useStateA(2); => stateA = 2;
const [stateB, setStateB] = useStateB(0); => stateB = 0;
const [stateC, setStateC] = useStateC(3); => stateC = 3;
```
<sup><del>Actually I'm not sure if it render twice when setStateA and setStateC is called at the same time. We should figure out how many.</del>I was right.</sup>

> Sometimes, It looks like `setState` working asyncronously. If you use `console.log` right after `setState` it isn't changed. Because you already assgined state before! So React look after you assgined value. **`setState` gives a signal for re-rendering. You can see the changed value after re-rendering.**

![스크린샷 2024-04-09 113031](https://github.com/honaseung/react-deep-dive/assets/53066347/8f59a9be-569e-44f4-8c85-bf077f12bca2)

### 3. Which one is good for make as state? (my opinion)
> **I think `state` is usually used to render. State usually kind of data for showing. But also it could be changed. If it's for showing, but no changing, it doesn't need to be `state`. You can just use `javascript` `variables` instead of `state`.**

![image](https://github.com/honaseung/react-deep-dive/assets/53066347/74fb4c6c-0cd1-4e96-a221-cd94ab71adf7)

#### lazy  assign
> Another thing I'd like to talk is lazy assign. If you give a default value for `state` as a `function`. This `funtion` is only one time called. Other states's change doesn't run this `function` once again. It's prevented for guaranting performance.

#### dependency array
> We talked about `useEffect` above. And I told you if you want to watch specific `state` then you should put it into '[]'. But you have to know this too. If you use some `state` or `props` inside of `useEffect`. Then it's better to place it into '[]'. If you don't, it will not run. Even it has been changed.

**Comment by honaseung** [comment]
당신이 클래스 컴포넌트에 친근하다면 'state'에 대해서 잘 알고 있을 것입니다. 'useState' 는 함수형 컴포넌트에서 'state' 를 만들어 주는 함수입니다. 그렇다면 'state' 란 무엇 일까요? 아래 3개의 질문에 대해서 대답하실 수 있으시다면 'state' 에 대해서 잘 알고 계신다는 뜻입니다.


왜 'state' 를 사용해야 하는가
'state' 는 어떻게 작동하는가
무엇이 'state' 가 되기에 적합한가
이 질문들은 굉장히 단순하고 기본적인 질문들입니다. 1번부터 설명하겠습니다.


리액트에서 리액트는 컴포넌트를 많이 렌더링합니다. 리랜더링이 발생한다는 것은 함수형 컴포넌트를 다시 만들기 위해 코드를 처음부터 시작해서 끝까지 다시 실행 한다는 의미입니다. 당신의 컴포넌트가 아래와 같다고 생각해봅시다.

const Test = () => {
  const a = 1;
  return (
    <div>TEST</div>
  );
};

리액트가 이것을 리랜더링을 할때 코드는 `const a = 1;` 에서부터 시작합니다. 이해하셨나요? 우리는 리랜더링하기전에 값을 변경하였다고 하더라도 값을 계속 유지할 수 없습니다. `a` 는 항상 `1` 입니다. 만약 당신이 함수형 컴포넌트가 변경된 값을 기억하길 바란다면 (또는 변경 이전 값) `useState` 를 사용해야 합니다.

2. How state work?

우리는 `state` 가 무엇인지 이야기 했습니다. 이번에는 이것이 어떻게 작동하는지 볼 것 입니다. 로우 레벨에서는 아마 아래와 같이 동작할 것입니다. (저의 추측입니다.)

At the raw level
.
.
.
At first render
const [stateA, setStateA] = useStateA(0); => stateA = 0;
At re-render after setStateA(1);
const [stateA, setStateA] = useStateA(1); => stateA = 1;

매 렌더링마다 새로운 변수를 할당하는 것과 같습니다. 매 렌더링은 서로 다른 `state` 값을 가질 수 있습니다. 렌더링이 일어나는 이유중 하나는 `setState` 입니다. `state` 값이 변경된다면, 리렌더링은 새로운 `state` 값과 함께 일어납니다. 만약 `state` 가 한개가 아닌 여러개라면 아래와 같을 것입니다.

At first render
const [stateA, setStateA] = useStateA(0); => stateA = 0;
const [stateB, setStateB] = useStateB(0); => stateB = 0;
const [stateC, setStateC] = useStateC(0); => stateC = 0;
At re-render after setStateA(1);
const [stateA, setStateA] = useStateA(1); => stateA = 1;
const [stateB, setStateB] = useStateB(0); => stateB = 0;
const [stateC, setStateC] = useStateC(0); => stateC = 0;
At re-render after setStateA(2) and setStateC(3);
const [stateA, setStateA] = useStateA(2); => stateA = 2;
const [stateB, setStateB] = useStateB(0); => stateB = 0;
const [stateC, setStateC] = useStateC(3); => stateC = 3;
Actually I'm not sure if it render twice when setStateA and setStateC is called at the same time. We should figure out how many.I was right.


때때로 `setState` 는 비동기로 작동하는 것처럼 보입니다. 당신이 `console.log` 를 `setState` 바로 뒤에 사용한다면 `state` 값은 변경되기 전의 값이 출력됩니다. 왜냐하면 당신은 이미 값을 할당하였고 할당한 값을 사용했기 때문입니다. `setState` 에 넘겨준 값은 다른 렌더링때 확인할 수 있습니다.

스크린샷 2024-04-09 113031

3. Which one is good for make as state? (my opinion)
I think state is usually used to render. State usually kind of data for showing. But also it could be changed. If it's for showing, but no changing, it doesn't need to be state. You can just use javascript variables instead of state.

image

More things to say
lazy assign
Another thing I'd like to talk is lazy assign. If you give a default value for state as a function. This funtion is only one time called. Other states's change doesn't run this function once again. It's prevented for guaranting performance.

dependency array
We talked about useEffect above. And I told you if you want to watch specific state then you should put it into '[]'. But you have to know this too. If you use some state or props inside of useEffect. Then it's better to place it into '[]'. If you don't, it will not run. Even it has been changed.

**Comment by chaminju927** [comment]
"raw lever"이 무엇을 의미하는건가요? 실행되는 날것의 코드...?

**Reply by honaseung** [reply]
raw level 이라는 것은 정식 용어는 아닙니다. 본의 아니게 혼란을 드렸네요. 여기서 말하는 raw level 이라는 것은 우리가 외부에서 사용하는 수준이 아닌 라이브러리 내부에서 제일 낮은 단계의 `javascript` 로 실행되는 단계를 의미하는 것입니다.

**Comment by chaminju927** [comment]
![image](https://github.com/honaseung/react-deep-dive/assets/118426671/31ffa3a4-1044-4a25-9cbe-5609c721da4c)
setStateA(1)이 실행된 후 리렌더링시 stateA의 초기값이 1이다라고 생각하는게 맞나요?

**Reply by honaseung** [reply]
네 그렇게 생각하셔도 무방합니다! 최초엔 stateA = 0 이었지만 setState(1) 로 리렌더링이 될 때에는 stateA = 1 로 초기화 되는 것이지요. 초기값이라는 기준이 최초 렌더링을 바라보고 하는 말이라면 잘못된 생각이지만요. 😂
